### PR TITLE
[WIP]Implement the master-worker framework

### PR DESCRIPTION
**Pull request by liuzix**
-  This PR contains a prototype implementation of the master-worker framework, and it can contain some bugs.

### CODE REVIEW

File: lib/worker.go [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+}
+
+type BaseWorker struct {
+	impl WorkerImpl
+
+	messageHandlerManager p2p.MessageHandlerManager
+	messageRouter         p2p.MessageRouter
+	metaKVClient          metadata.MetaKV
+
+	master *masterManager
```
Comment (nongfushanquan) [inline lib/worker.go:55]: master client?

File: lib/worker.go [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+		case <-ticker.C:
+			status, err := w.impl.Status()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			workload, err := w.impl.Workload()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if err := w.master.SendStatus(ctx, status, workload); err != nil {
```
Comment (nongfushanquan) [inline lib/worker.go:182]: executor is expected to send workload to the resource manager

File: lib/worker_manager.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	id := info.ID
+	_, exists := m.workerInfos[id]
+	return !exists
+}
+
+func (m *workerManager) onWorkerCreated(id WorkerID, exeuctorNodeID p2p.NodeID) {
+	m.putWorkerInfo(&WorkerInfo{
```
Comment (nongfushanquan) [inline lib/worker_manager.go:261]: error  should be handled here

File: lib/worker_manager.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+	}
+	return value, true
+}
+
+func (m *workerManager) putWorkerInfo(info *WorkerInfo) bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	id := info.ID
+	_, exists := m.workerInfos[id]
```
Comment (nongfushanquan) [inline lib/worker_manager.go:256]: why not add info to the workerInfos[]

File: lib/master.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+	}
+
+	if err := m.markInitializedInMetadata(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (m *BaseMaster) Poll(ctx context.Context) error {
+	OfflinedWorkers := m.workers.Tick(ctx, m.messageRouter)
```
Comment (nongfushanquan) [inline lib/master.go:126]: the workers.Tick() should be called in the init()?,  the Poll will make the heartbeat timeout

File: lib/worker_manager.go [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+	_, err := messageClient.TrySendMessage(ctx, prefixedTopic, message)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (w *workerHandleImpl) Status() *WorkerStatus {
+	w.manager.mu.Lock()
+	defer w.manager.mu.Unlock()
```
Comment (amyangfei) [inline lib/worker_manager.go:68]: Can use a `RWMutex`

File: lib/worker_manager.go [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+
+func (m *workerManager) UpdateWorkload(msg *WorkloadReportMessage) {
+	info, ok := m.getWorkerInfo(msg.WorkerID)
+	if !ok {
+		log.L().Info("received workload update for non-existing worker",
+			zap.String("master-id", string(m.masterID)),
+			zap.Any("msg", msg))
+		return
+	}
+	info.workload = msg.Workload
```
Comment (amyangfei) [inline lib/worker_manager.go:220]: The lock used in these two functions are not same, in `HandleHeartBeat`, `m.mu` protects worker info getter and worker info update. in `UpdateWorkload`, `m.mu` protects worker info getter only.

File: lib/master.go [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+			Cmd: client.CmdDispatchTask,
+			Req: &pb.DispatchTaskRequest{
+				TaskTypeId: int64(workerType),
+				TaskConfig: config,
+			},
+		})
+		if err != nil {
+			err1 := m.OnWorkerDispatched(nil, errors.Trace(err))
+			if err1 != nil {
+				m.OnError(errors.Trace(err))
```
Comment (amyangfei) [inline lib/master.go:324]: Should we use `err` or `err1` here?

File: lib/master.go [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+			}
+			return
+		}
+		dispatchTaskResp := executorResp.Resp.(*pb.DispatchTaskResponse)
+		errCode := dispatchTaskResp.GetErrorCode()
+		if errCode != pb.DispatchTaskErrorCode_OK {
+			err1 := m.OnWorkerDispatched(
+				nil, errors.Errorf("dispatch worker failed with error code: %d", errCode))
+			if err1 != nil {
+				m.OnError(errors.Trace(err))
```
Comment (amyangfei) [inline lib/master.go:334]: ditto

File: lib/worker.go [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+}
+
+type BaseWorker struct {
+	impl WorkerImpl
+
+	messageHandlerManager p2p.MessageHandlerManager
+	messageRouter         p2p.MessageRouter
+	metaKVClient          metadata.MetaKV
+
+	masterClient *masterClient
```
Comment (amyangfei) [inline lib/worker.go:54]: Is `jobMasterClient` better

**Review by amyangfei — APPROVED** [review]

**Pull request merged**